Sort emoji order based on search string.

tejaskasetty opened the pull request:
Follows similar sort order present in  emoji typeahead.
Fixes #4806

smarx commented:
*Automated message from Dropbox CLA bot*

@tejaskasetty, it looks like you've already signed the Dropbox CLA. Thanks!

HarshitOnGitHub commented:
@tejaskasetty Please take care of the lint errors.

tejaskasetty commented:
@HarshitOnGitHub  Fixed the lint errors.

timabbott commented:
I merged the first commit; for the second, I think you should remove the code duplication between `reactions.js` and `emoji_picker.js` of things like `promote_popular`.

timabbott commented:
Also, playing around with this, it feels much more laggy than before (it takes a long time to update after you type a few characters); probably you should use a different method for updating the list of elements.  The Chrome profiler may be helpful in this.

tejaskasetty commented:
Okay.

tejaskasetty commented:
@timabbott The lag might be because of the setTimeout with time parameter set to (300ms). This i did to avoid multiple filters while entering search string. Eg. If the user wanted to search "hamb" usually it used to do 5 searches for  h, ha,..., hamb (unnecessary computation). instead if the user is typing fast I thought we can skip this and directly filter when the user pauses or gives a  considerable gap before continuing the input. i can reduce this gap to make the search happen faster after the pause. The average typing speed is considered to be 40 rpm around 190 characters per minute. So thought 350 ms per character. I can reduce this to 250 ms. But will check if there is better approach.
And for the empty string i thought i would use the same promote_popular to get the order same as the inital order when the emoji-reaction-picker is opened. Shall i make the implementation in emoji_picker accessible in other modules.

tejaskasetty commented:
@timabbott
I have removed the code duplication 
I have made few optimizations with the way the emoji's are order. With the help of javascript profiler I was able to assess the performance of filter_emoji and right now its seems to be near to the previous implementation but would request you to check and give a feed back on the user experience.

tejaskasetty commented:
@HarshitOnGitHub Would be nice if you could assess and give a feed back on how the user experience would be, if it is still laggy.

timabbott commented:
There's still a noticable lag between when you type a character in the emoji picker and the list updates in this PR, that isn't present on chat.zulip.org.

tejaskasetty commented:
@timabbott I have further optimized it. This seems to be good and satisfying. Please checkout it out.

timabbott commented on static/js/emoji_picker.js:
```diff
@@ -154,28 +159,54 @@ function get_selected_emoji() {
+}
+
+function sort_emoji_order(emoji_list, query) {
+    set_emoji_order(emoji_list, true);
```
why does this get called even if there is a nonempty query?

timabbott commented on static/js/emoji_picker.js:
```diff
@@ -154,28 +159,54 @@ function get_selected_emoji() {
     }).toArray();
 };
 
+function set_emoji_order(emoji_list, is_set_default) {
```
This function, and the one after, could use either some comments explaining the reasoning.  Or maybe better names?

tejaskasetty replied:
Okay.

tejaskasetty commented:
@timabbott 
 Changed the function names. Added comments explaining briefly what each part in the two functions does.
About the reaction hotkeys not working. Can tell me how to reproduce it? Because its working fine when I try

timabbott commented:
The reaction hotkeys work, unless you type a character in the filter box.  You should be able to type a few characters in the filter box and then hit enter/down/etc.

timabbott commented:
(BTW, asking for help reproducing issues with your PR on chat.zulip.org can be a good approach, rather than waiting for me to get to it).

tejaskasetty commented:
Oh Okay. By the way I had already fixed the issue with the hotkeys and reported it in code review.

timabbott commented:
Merged, thanks for all your work on this @tejaskasetty!  @HarshitOnGitHub FYI.

@brockwhittaker this may conflict with your categories work on the emoji picker, but I don't see a PR yet; my hope is that this is changing different parts of the picker code.

As a sidenote, I think this would be a good module to target next for node unit tests and refactoring; I feel like `emoji_picker.js` is far harder to read than it should be.
